### DISCUSSION TITLE
`ADR Suggestion` Release process for EasyScience

### DISCUSSION THREAD

**Discussion by rozyczko** [easyscience/.github#28] [Ideas]
Similar to what we have for EasyDiffractionApp ( https://github.com/EasyScience/EasyDiffractionApp/discussions/247 ) we should document the new release checklist. (Note that the EDA checklist is already out of date)

It will probably be difficult to have the same checklist for the core, the libraries, and the apps, but some parts should be identical throughout all the projects.

Currently, the release in EasyScience goes like this:

```mermaid
graph TD;
Develop --> Pre-release;
Pre-release --> Master;
```

We need the `Pre-release` branch since the `release_drafter` workflow works on this branch only, so we either accept this step, modify scripts, or create release notes by hand.

The proposed checklist would be:

**DEVELOP BRANCH**
- [ ] Update `src\easyscience\__version__.py` with the new release version string
- [ ] Merge develop into `pre-release` branch, with **NO SQUASH** strategy (all commit info preserved)

**PRE-RELEASE BRANCH**
- [ ] Make a PR for `pre-release branch` -> `master` merge, with **SQUASH** strategy (all commits merged into one)
- [ ] Assure the build passes

**MASTER BRANCH**
- [ ] Merge `pre-release` into `master`
- [ ] Draft release on GH: check the release notes and release draft generated in `pre-release`
- [ ] Tag the release accordingly

**Comment by andped10** [comment]
This comment only concerns when to do merge with or without SQUASH.

# develop (SQUASH)
I think that we should always SQUASH when doing merges into `develop`.  Merges into `develop` will either be the inclusion of a feature branch or a bug fix.  For such branches there often are numerous commits to get to the point where the PR was accepted and the new code can be merged in.  These commits will in general not be of any interest (in case one wants to have a closer look this can be done by recreating the branch that was merged).  To get rid of the "noisy" commits there were made on the branch it should be merged using SQUASH. The result should be a `develop` branch where each commit (merged PR) either corresponds to the implementation of a feature or the fix of a bug.

# pre-release (Not SQUASH)
To progress the `pre-release` branch all new features / bug fixed from `develop` should be merged in.  In this case the merge should NOT SQUASH because we want to keep the history of the individual commits (features / bug fixes).

# master (Not SQUASH)
The same goes when merging to `master`.  In this case we also want to keep the history of the features and bugs and the merge should NOT SQUASH these into a single commit.  Furthermore, when keeping the commits it should also be possible to git to resolve the real changes that have happen when comparing the ´develop´, ´pre-release´ and ´master´ branches

**Comment by andped10** [comment]
I seem to remember we had a discussion where we decided to skip the pre-release step.  However, I'm not able to find the discussion :(

**Reply by rozyczko** [reply]
We did. https://github.com/orgs/EasyScience/discussions/4

This is why I mentioned that if we decide to skip the `pre-release` branch, we'd have to modify the release notes action to act on master commits (?) not pre-release

**Comment by andped10** [comment]
Is this discussion not relevant for all our `EasyScience` repos and not just `EasyScience/EasyScience`

**Reply by rozyczko** [reply]
It probably should start here, since we want a proper process for `EasyScience`. Then, let's compare those checklists from diffraction lib/app and reflectometry lib/app and see how much we can streamline all of them.

**Comment by andped10** [comment]
For reflectometry there wiki pages for the release checklist:
Lib: https://github.com/EasyScience/EasyReflectometryLib/wiki/Release-checklist 
App: https://github.com/EasyScience/EasyReflectometryApp/wiki/Release-checklist

**Comment by rozyczko** [comment]
Then how about:
 
 ```mermaid
 graph TD;
 Develop --> Master;
 ```

The checklist:

**DEVELOP BRANCH**

* [ ]  Update `src\easyscience\__version__.py` with the new release version string
* [ ] Generate release notes

 **MASTER BRANCH**

* [ ]  Merge `develop` into `master`
* [ ]  Draft release on GH: check the release notes and release draft
* [ ]  Tag the release accordingly
* [ ]  Assure the artifacts are uploaded to pypi

**Reply by andped10** [reply]
I assume it should be: Merge `develop` into `master`

I would love if we could fetch the tag info in the `pyproject.toml` file and then use that as the version number.  However, until we reach that point it is fine to use `__version__.py`, but we should the remove all traces of version in `pyproject.toml` or fetch the info from the version file.

Are we not able to use the `release-drafter` functionality by pointing the pipeline to `master` rather than `pre-release`?

**Reply by rozyczko** [reply]
Fixed the 1st typo.

Agreed with use of release-drafter on master. It would be great to put it there and have the notes generated automatically.

Requirement to have to set the version # in one place only is good. I am not sure whether `pyproject.toml` can be queried in the GH workflow, so using `__version__.py` is a cop out. If there's a method to pick the version from toml - we should use it.

**Reply by AndrewSazonov** [reply]
What is the advantage of storing the release version in a separate file `__version__.py` instead of `pyproject.toml`? Is it easy to extract the version form `__version__.py` in a GH workflow?

I prefer to store all the project information in one place, i.e. in `pyproject.toml`. The hatch build system is able to properly extract the static version from `pyproject.toml`.

I don't know if there is an easy way to extract this version from `pyproject.toml` in a GH workflow. One could have something like `utils.py` with a simple function to parse toml, using, for example, the `tomllib` library standard for Python since 3.11. A similar approach is currently used in EasyDiffractionBeta. Or you could try calling python with the '-c' option in the CI build script to extract the version, e.g. like this 

```
python3 -c 'import tomllib; f = open("pyproject.toml", "rb"); print(tomllib.load(f)["project"]["version"]); f.close()'
```

It's a bit long, but should work.

**Reply by damskii9992** [reply]
Is this checklist still valid @rozyczko? In that case I would like to convert this ADR suggestion into a proper ADR for the release proces.

**Reply by rozyczko** [reply]
I think many of the points have been looked at and modified by the recent work on templating the process, by @AndrewSazonov

**Reply by rozyczko** [reply]
The diagram underneath is still valid and should be in our Process Bible

**Comment by andped10** [comment]
I have made a more detailed diagram of the merging strategy.  Could you please take a look and let me know if you understand what is going on / what needs to be cleared up:
![git-merging](https://github.com/user-attachments/assets/ea59f696-ce0c-4e9c-9aa4-60138de18514)

**Reply by rozyczko** [reply]
Additionally, we should agree on dependency chain for multi-module PRs.

When merging to `develop`, the feature branch should already depend ONLY on master (released) versions of local dependendant modules.

**Reply by damskii9992** [reply]
For context, the reason for this decision is that otherwise merging feature branches will risk breaking develop branches of modules further downstream when merging to develop.

**Comment by damskii9992** [comment]
Since the points in this discussion relates mostly to organization-level stuff such as branching and merging-strategies, and since the repository-specific release process points are no longer valid, as the release processes are currently being streamlined with copier templates, which will be covered in a future ADR, this discussion will be moved to the organization level and closed.
